src/MatrixNetwork.jl

```diff
@@ -28,32 +28,53 @@ function MatrixNetwork(ei::Vector{Int64},ej::Vector{Int64},n::Int64)
 end
 
 
-function _matrix_network_direct{T}(A::SparseMatrixCSC{T,Int64})
+function _matrix_network_direct(A::SparseMatrixCSC{T,Int64}) where T
     return MatrixNetwork(size(A,2),A.colptr,A.rowval,A.nzval)
 end
 
-function _matrix_network_direct{T}(A::SparseMatrixCSC{T,Int64},v)
+function _matrix_network_direct(A::SparseMatrixCSC{T,Int64},v) where T
     nzval = ones(typeof(v),length(A.nzval))
     return MatrixNetwork(size(A,2),A.colptr,A.rowval,nzval)
 end
 
 
-import Base.sparse, Base.size, Base.*, Base.A_mul_B!, Base.At_mul_B, Base.At_mul_B!
+import SparseArrays.sparse, Base.size, Base.*, LinearAlgebra.mul!, Base.convert, Base.adjoint
 
 """
 Return back an adjacency matrix representation
 of the transpose. This requires no work. 
 """
-function sparse_transpose{T}(A::MatrixNetwork{T})
+function sparse_transpose(A::MatrixNetwork{T}) where T
     return SparseMatrixCSC(A.n,A.n,A.rp,A.ci,A.vals)
 end
 
+# function adjoint(A::MatrixNetwork{T}) where T
+#     return SparseMatrixCSC(A.n,A.n,A.rp,A.ci,A.vals)
+# end
+# function adjoint(A::MatrixNetwork{T}) where T
+#     SparseMatrixCSC(A.n,A.n,A.rp,A.ci,A.vals)
+#     return
+# end
+function adjoint(A::MatrixNetwork{T}) where T
```

Comment:
This came up because we want to support `M'*x` -- and now there's only the `mul!` function.
The current state mimics Line 87 of the below old code, but maybe `adjoint` should be kept internal in this case.

This is the old code: https://github.com/nassarhuda/MatrixNetworks.jl/blob/193d486d00f17f8f3bddd29ae8ae9b8e7f7310a5/src/MatrixNetwork.jl#L76-L87

There's an alternative and that is to create a new type and return it in the adjoint function but that will only serve the purpose of `M'*x` so I'm not sure having a new type was the right thing here.

Or alternatively returning an Adjoint{T,SparseMatrixCSC} in the `adjoint` function, but that will require forming the underlying sparse matrix and calling adjoint on it.

So, I went for the first option (mimicking the old behavior in the sense that `M'x` will boil down to something equivalent to line 86 above).

Reply:
Ah, how annoying...

* to be consistent, adjoint / transpose should be cheap and lazy...
* But then we have to define a few things for Adjoint{MatrixNetwork}, right?
* So let's keep it consistent with the sparse interface:

mul!(Adjoint{MatrixNetwork},...) ... does what your current line does...
copy(Adjoint{MatrixNetwork},...) returns a copy with the real transpose in place.
sparse(Adjoint{MatrixNetwork},...) does the same as copy.

Make sense?

Reply:
Check changes here e78c50bb1ca0413f5207c6f904ee066672d4ee99